src/ImageWrapper.cpp

```diff
@@ -0,0 +1,160 @@
+  {
+    Gosu::Bitmap bitmap = Gosu::layout_text(text, font, font_height, spacing, width,
+                                              (Gosu::Alignment)align, font_flags);
+    return reinterpret_cast<Gosu_Image*>( new Gosu::Image(bitmap, image_flags) );
+  }
+
+  Gosu_Image *Gosu_Image_create_from_blob(unsigned char* blob, int byte_count, int columns, int rows, unsigned image_flags)
+  {
+    std::size_t size = columns * rows * 4;
+    Gosu::Bitmap bitmap(columns, rows, Gosu::Color::NONE);
+
+    if (byte_count == size) {
+      // 32 bit per pixel, assume R8G8B8A8
+      std::memcpy(bitmap.data(), blob, size);
+    }
+    else if (byte_count == size * sizeof(float)) {
+      // 128 bit per channel, assume float/float/float/float
+      const float *in = reinterpret_cast<const float *>( blob );
+      Gosu::Color::Channel *out = reinterpret_cast<Gosu::Color::Channel *>(bitmap.data());
+      for (int i = size; i > 0; --i) {
+        *(out++) = static_cast<Gosu::Color::Channel>(*(in++) * 255);
+      }
+    }
+    else {
+      return nullptr; // abort?
+    }
+
+    return reinterpret_cast<Gosu_Image *>(new Gosu::Image(bitmap, image_flags));
+  }
+
+  Gosu_Image *Gosu_Image_create_from_subimage(Gosu_Image *image, int left, int top, int width, int height)
+  {
+    Gosu::Image *gosu_image = reinterpret_cast<Gosu::Image *>(image);
+    std::unique_ptr<Gosu::ImageData> image_data = gosu_image->data().subimage(left, top, width, height);
+
+    return reinterpret_cast<Gosu_Image *>(image_data.get() ? new Gosu::Image(std::move(image_data)) : nullptr);
+  }
+
+  void Gosu_Image_create_from_tiles(const char* source, int tile_width, int tile_height, void function(void* data, Gosu_Image* image), void* data, unsigned image_flags)
+  {
+    std::vector<Gosu::Image> gosu_images = Gosu::load_tiles(source, tile_width, tile_height, image_flags);
+
+    for (Gosu::Image &img : gosu_images) {
+      function(data, reinterpret_cast<Gosu_Image*>(new Gosu::Image(img) ));
+    }
+  }
+
+  // Properties
+  unsigned Gosu_Image_width(Gosu_Image* image)
+  {
+    return reinterpret_cast<Gosu::Image*>( image )->width();
+  }
+
+  unsigned Gosu_Image_height(Gosu_Image* image)
+  {
+    return reinterpret_cast<Gosu::Image*>( image )->height();
+  }
+
+  Gosu_GLTexInfo* Gosu_Image_gl_tex_info(Gosu_Image* image)
+  {
+    Gosu_GLTexInfo* tex_info;
+    // FIXME: accessing gosu_texture_info causes segfault
+    const Gosu::GLTexInfo* gosu_texture_info = reinterpret_cast<Gosu::Image*>( image )->data().gl_tex_info();
+    if (gosu_texture_info) {
+      tex_info->texture_name = gosu_texture_info->tex_name;
+      tex_info->left         = gosu_texture_info->left;
+      tex_info->right        = gosu_texture_info->right;
+      tex_info->top          = gosu_texture_info->top;
+      tex_info->bottom       = gosu_texture_info->bottom;
+      return tex_info;
+    }
```

Comment:
Can this simply return a pointer to a `Gosu::GLTexInfo` struct cast as a `Gosu_GLTexInfo`, since they share the same layout?

```C
Gosu::Image gosu_image = reinterpret_cast<Gosu::Image*>( image );
Gosu_GLTexInfo* texture_info = reinterpret_cast<Gosu_GLTexInfo*>( new Gosu::GLTexInfo(gosu_image->data().gl_tex_info()) );

return texture_info;
```

Reply:
Should work, even if one is defined inside `extern "C"` and the other is C++. Let's go with the simple solution first.